### PR TITLE
Sharing: allow site owners to specify a custom Facebook app ID.

### PR DESCRIPTION
**Pull request by jeherve**
This custom App ID will be used with the official Facebook Share button,
with the Facebook Like Box, and with the Facebook shortcode.

It also avoids having the Facebook Share button display a link to WordPress.com by default. This behaviour was introduced in 82cbc63e5bddeb0a9a261d769e3c579180f591b5, but an App ID isn't required for that button:
https://developers.facebook.com/docs/plugins/share-button

Site owners can then specify a custom App ID with a code snippet like this one:

```php
function jeherve_custom_fb_app_id() {
	return '190102077773016';
}
add_filter( 'jetpack_sharing_facebook_app_id', 'jeherve_custom_fb_app_id' );
```

Originally reported here:
https://wordpress.org/support/topic/how-to-remove-link-to-wordpresscom-on-facebook-shares?replies=3&view=all#post-7332376

### CODE REVIEW

File: modules/sharedaddy/sharing-sources.php [hunk 1]
```diff
@@ -724,7 +724,18 @@ public function display_footer() {
...
 			}
-			?><div id="fb-root"></div><script>(function(d, s, id) { var js, fjs = d.getElementsByTagName(s)[0]; if (d.getElementById(id)) return; js = d.createElement(s); js.id = id; js.src = '//connect.facebook.net/<?php echo $locale; ?>/sdk.js#xfbml=1&appId=249643311490&version=v2.3'; fjs.parentNode.insertBefore(js, fjs); }(document, 'script', 'facebook-jssdk'));</script><?php
+			/**
+			 * Filter the App ID used in the official Facebook Share button.
+			 *
+			 * @since 3.8.0
+			 *
+			 * @param int $fb_app_id Facebook App ID. Default to empty.
+			 */
+			$fb_app_id = absint( apply_filters( 'jetpack_sharing_facebook_app_id', '' ) );
```
Comment (dereksmart) [inline modules/sharedaddy/sharing-sources.php]: Let's continue to default to the WP app ID, since this has to be synced back there.

Comment (jeherve) [inline modules/sharedaddy/sharing-sources.php]: Ok! aba5ce739875c92388a83af27411fc51cdcfb64d

File: modules/shortcodes/facebook.php [hunk 2]
```diff
@@ -40,6 +40,15 @@ function jetpack_facebook_embed_handler( $matches, $attr, $url ) {
 		return $embed . '<script src="//connect.facebook.net/en_US/all.js#xfbml=1"></script>';
 	} else {
 		wp_enqueue_script( 'jetpack-facebook-embed', plugins_url( 'js/facebook.js', __FILE__ ), array( 'jquery' ), null, true );
+		/** This filter is documented in modules/sharedaddy/sharing-sources.php */
+		$fb_app_id = absint( apply_filters( 'jetpack_sharing_facebook_app_id', '249643311490' ) );
+		wp_localize_script(
+			'jetpack-facebook-embed',
+			'jpfbembed',
+			array(
+				'appid' => $fb_app_id
```
Comment (aduth) [inline modules/shortcodes/facebook.php]: Some odd tabbing here.

File: modules/sharedaddy/sharing-sources.php [hunk 3]
```diff
@@ -738,9 +738,19 @@ public function display_footer() {
...
-			<div id="fb-root"></div>
-			<script>(function(d, s, id) { var js, fjs = d.getElementsByTagName(s)[0]; if (d.getElementById(id)) return; js = d.createElement(s); js.id = id; js.src = '//connect.facebook.net/<?php echo $locale; ?>/sdk.js#xfbml=1&appId=249643311490&version=v2.3'; fjs.parentNode.insertBefore(js, fjs); }(document, 'script', 'facebook-jssdk'));</script>
+			/**
+			 * Filter the App ID used in the official Facebook Share button.
+			 *
+			 * @since 3.8.0
+			 *
+			 * @param int $fb_app_id Facebook App ID. Default to empty.
+			 */
+			$fb_app_id = absint( apply_filters( 'jetpack_sharing_facebook_app_id', '249643311490' ) );
```
Comment (aduth) [inline modules/sharedaddy/sharing-sources.php]: Is calling `absint` going to be a problem with large app IDs? For example, my own example app ID is "1485813205022947", which is larger than `PHP_INT_MAX` (`2147483647`) on 32-bit devices. Maybe instead, we verify numeric value using [`is_numeric`](http://php.net/manual/en/function.is-numeric.php), which can support any number of digits.

File: modules/shortcodes/facebook.php (hunk 2 above)
Comment (dereksmart) [inline modules/shortcodes/facebook.php]: fixed

File: modules/sharedaddy/sharing-sources.php [hunk 4]
```diff
@@ -738,9 +738,21 @@ public function display_footer() {
...
 			}
-			?>
-			<div id="fb-root"></div>
-			<script>(function(d, s, id) { var js, fjs = d.getElementsByTagName(s)[0]; if (d.getElementById(id)) return; js = d.createElement(s); js.id = id; js.src = '//connect.facebook.net/<?php echo $locale; ?>/sdk.js#xfbml=1&appId=249643311490&version=v2.3'; fjs.parentNode.insertBefore(js, fjs); }(document, 'script', 'facebook-jssdk'));</script>
+			/**
+			 * Filter the App ID used in the official Facebook Share button.
+			 *
+			 * @since 3.8.0
+			 *
+			 * @param int $fb_app_id Facebook App ID. Default to empty.
```
Comment (aduth) [inline modules/sharedaddy/sharing-sources.php]: The comment here isn't accurate - it defaults to the WordPress.com app ID.

File: modules/shortcodes/facebook.php (hunk 2 above)
Comment (aduth) [inline modules/shortcodes/facebook.php]: Not sure if Facebook will barf on an empty (non-numeric) app ID, but I suppose that's partly on the developer to ensure a valid ID is passed.

**Comment by aduth** [comment]
One minor nitpick on filter commenting, but this looks and works well in my testing :+1:

**Pull request merged**